tutorials/optimization/gpu_optimization.rst

```diff
@@ -0,0 +1,258 @@
+far away (so it looks tiny), use a smaller level of detail (LOD) model.
+
+Pay attention to the additional vertex processing required when using:
+
+-  Skinning (skeletal animation)
+-  Morphs (shape keys)
+-  Vertex-lit objects (common on mobile)
+
+Pixel / fragment shaders - fill rate
+====================================
+
+In contrast to vertex processing, the costs of fragment shading has increased
+dramatically over the years. Screen resolutions have increased (the area of a 4K
+screen is ``8,294,400`` pixels, versus ``307,200`` for an old ``640x480`` VGA
+screen, that is 27x the area), but also the complexity of fragment shaders has
+exploded. Physically based rendering requires complex calculations for each
+fragment.
+
+You can test whether a project is fill rate limited quite easily. Turn off vsync
+to prevent capping the frames per second, then compare the frames per second
+when running with a large window, to running with a postage stamp sized window.
+Usually you will find the fps increases quite a bit using a small window, which
+indicates you are to some extent fill rate limited. If on the other hand there
+is little to no increase in fps, then your bottleneck lies elsewhere.
+
+You can increase performance in a fill rate limited project by reducing the
+amount of work the GPU has to do. You can do this by simplifying the shader
+(perhaps turn off expensive options if you are using a :ref:`SpatialMaterial
+<class_SpatialMaterial>`), or reducing the number and size of textures used.
+
+Consider shipping simpler shaders for mobile.
+
+Reading textures
+~~~~~~~~~~~~~~~~
+
+The other factor in fragment shaders is the cost of reading textures. Reading
+textures is an expensive operation (especially reading from several in a single
+fragment shader), and also consider the filtering may add expense to this
+(trilinear filtering between mipmaps, and averaging). Reading textures is also
+expensive in power terms, which is a big issue on mobiles.
+
+Texture compression
+~~~~~~~~~~~~~~~~~~~
+
+Godot compresses textures of 3D models when imported (VRAM compression) by
+default. Video RAM compression is not as efficient in size as PNG or JPG when
+stored, but increases performance enormously when drawing.
+
+This is because the main goal of texture compression is bandwidth reduction
+between memory and the GPU.
+
+In 3D, the shapes of objects depend more on the geometry than the texture, so
+compression is generally not noticeable. In 2D, compression depends more on
+shapes inside the textures, so the artifacts resulting from 2D compression are
+more noticeable.
+
+As a warning, most Android devices do not support texture compression of
+textures with transparency (only opaque), so keep this in mind.
+
+Post processing / shadows
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Post processing effects and shadows can also be expensive in terms of fragment
+shading activity. Always test the impact of these on different hardware.
+
```

Comment:
Add to end: Reducing the size of shadow maps can increase performance, both in terms of writing, and reading the maps.